### PR TITLE
GH-94851: check refcnt of immortal objects after finalization

### PR DESCRIPTION
**Pull request by kumaraditya303**
* Issue: gh-94851

Automerge-Triggered-By: GH:ericsnowcurrently

### CODE REVIEW

File: Tools/scripts/generate_global_objects.py [hunk 1]
```diff
@@ -293,6 +302,15 @@ def generate_runtime_init(identifiers, strings):
                 # since iter_files() ignores .h files.
                 printer.write(f'string = &_Py_ID({i});')
                 printer.write(f'PyUnicode_InternInPlace(&string);')
+        printer.write('')
+        printer.write('#ifdef Py_DEBUG')
+        printer.write("static inline void")
+        with printer.block("_PyStaticObjects_CheckRefcnt(void)"):
+            for i in immortal_objects:
+                with printer.block(f'if (Py_REFCNT({i}) < 999999999)', ';'):
```
Comment (ericsnowcurrently) [inline Tools/scripts/generate_global_objects.py]: Hard-coding the specific refcount isn't ideal.  Consider adding a constant to Include/internal/pycore_object.h.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Other than one small thing, LGTM.  I suspect we will rip this out once (if) PEP 683 lands.  Even then this is a helpful stopgap.

File: Tools/scripts/generate_global_objects.py (hunk 1 above)
Comment (kumaraditya303) [reply Tools/scripts/generate_global_objects.py]: Added `_PyObject_IMMORTAL_REFCNT` constant.

File: Tools/scripts/generate_global_objects.py [hunk 2]
```diff
@@ -293,6 +302,16 @@ def generate_runtime_init(identifiers, strings):
                 # since iter_files() ignores .h files.
                 printer.write(f'string = &_Py_ID({i});')
                 printer.write(f'PyUnicode_InternInPlace(&string);')
+        printer.write('')
+        printer.write('#ifdef Py_DEBUG')
+        printer.write("static inline void")
+        with printer.block("_PyStaticObjects_CheckRefcnt(void)"):
+            for i in immortal_objects:
+                with printer.block(f'if (Py_REFCNT({i}) < _PyObject_IMMORTAL_REFCNT)', ';'):
```
Comment (tiran) [inline Tools/scripts/generate_global_objects.py:310]: Why ``<``? Shouldn't the refcount be exactly the same?

Comment (kumaraditya303) [reply Tools/scripts/generate_global_objects.py:310]: Refcount can be higher if static types are used as they hold interned strings ex: `__add__`  and are never finalized so we only check for `<`.

File: Include/internal/pycore_object.h [hunk 3]
```diff
@@ -14,9 +14,11 @@ extern "C" {
 #include "pycore_pystate.h"       // _PyInterpreterState_GET()
 #include "pycore_runtime.h"       // _PyRuntime
 
+#define _PyObject_IMMORTAL_REFCNT 999999999
```
Comment (ericsnowcurrently) [inline Include/internal/pycore_object.h:20]: I recommend a comment here so no one gets confused about the word "immortal".  Perhaps something like this:

```suggestion
/* This value provides *effective* immortality, meaning the object should never
    be deallocated (until runtime finalization).  See PEP 683 for more details about
    immortality, as well as a proposed mechanism for proper immortality. */
#define _PyObject_IMMORTAL_REFCNT 999999999
```

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Thanks for adding the constant!  There's a comment to add and then this should be ready to merge.

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

**Comment by kumaraditya303** [comment]
Thanks for the reviews @ericsnowcurrently!

**Pull request merged**